xds/internal/xdsclient/authority.go

```diff
@@ -416,6 +500,76 @@ func (a *authority) handleADSResourceDoesNotExist(rType xdsresource.Type, resour
 	}
 }
 
+// handleRevertingToPrimaryOnUpdate is called when a resource update is received
+// from a server that is not the current active server. This method ensures that
+// all lower priority servers are closed and the active server is reverted to
+// the highest priority server that has sent an update.
+//
+// This method is only executed in the context of a serializer callback.
+func (a *authority) handleRevertingToPrimaryOnUpdate(serverConfig *bootstrap.ServerConfig) {
+	if a.activeXDSChannel != nil && a.activeXDSChannel.sc.Equal(serverConfig) {
+		// If the resource update is from the current active server, nothing
+		// needs to be done from fallback point of view.
+		return
+	}
+
+	if a.logger.V(2) {
+		a.logger.Infof("Received update from non-active server %q", serverConfig)
+	}
+
+	// If the resource update is not from the current active server, it means
+	// that we have received an update from a higher priority server and we need
+	// to revert back to it. This method guarantees that when an update is
+	// received from a server, all lower priority servers are closed.
+	serverIdx := 0
+	for _, cfg := range a.xdsChannelConfigs {
+		if cfg.sc.Equal(serverConfig) {
+			break
+		}
+		serverIdx++
+	}
+	if serverIdx == len(a.xdsChannelConfigs) {
+		// This can never happen.
+		a.logger.Errorf("Received update from an unknown server: %v", serverConfig)
+		return
+	}
+	a.activeXDSChannel = a.xdsChannelConfigs[serverIdx]
+
+	// Close all lower priority channel.
+	for i := serverIdx + 1; i < len(a.xdsChannelConfigs); i++ {
+		cfg := a.xdsChannelConfigs[i]
+
+		// Unsubscribe any resources that were subscribed to, on this channel
+		// and remove it from the resource cache. When a ref to a channel is
+		// being released, there should be no more references to it from the
+		// resource cache.
+		for rType, rState := range a.resources {
+			for resourceName, state := range rState {
+				idx := 0
+				for _, xc := range state.xdsChannelConfigs {
+					if xc != cfg {
+						state.xdsChannelConfigs[idx] = xc
+						idx++
+						continue
+					}
+					xc.xc.unsubscribe(rType, resourceName)
+				}
+				state.xdsChannelConfigs = state.xdsChannelConfigs[:idx]
```

Comment:
I've spent 10 minutes staring at this and I can't really figure out what's going on.

Maybe write some comments and say what idx is.

This continue logic feels a little funky as well. You loop through all the xDS Channel configs without breaking, and iterate the index except only when xc = cfg, then you unsubscribe. Then you take the slice up to the len(xc) - 1 (you loop through unconditionally) which will only throw away the last config in the slice of xDS Channel Configs? That seems wrong?

Also I find some of these go style short variable names confusing. I can't keep track of what each one is semantically and what each one represents.